torch_xla/csrc/torch_util.cpp

```diff
@@ -11,10 +12,13 @@ namespace torch_xla {
 void SymIntElements::AddSymIntNodeElements(c10::SymInt& size) {
   if (size.is_symbolic()) {
     // c10::SymInt --(convert)--> c10::SymIntNode --(cast)-->
-    // lazy::SymIntNodeImpl
+    // PythonSymNodeImpl --> getPyObj() --> XLASymIntNodeImpl
     // --(get)--> lazy::NodePtr --(cast)--> lazy::DimensionNode
     c10::SymNode symbolicIntNode = size.toSymNodeImpl();
-    auto* lazySymNode = dynamic_cast<XLASymNodeImpl*>(symbolicIntNode.get());
+    auto* pySymNode =
+        dynamic_cast<torch::impl::PythonSymNodeImpl*>(symbolicIntNode.get());
+    auto symNode = py::cast<c10::SymNode>(py::handle(pySymNode->getPyObj()));
```

Comment:
Would you mind explaining a bit what this part `py::cast<c10::SymNode>(py::handle(pySymNode->getPyObj()))` is doing?

Additionally, since this file is completely in c++, I wonder why there is python object involved.

Reply:
As I mentioned in the comment above, `PythonSymNodeImpl` is a container for python `SymNode`. In case of XLA, that python object is created by pybind when we return a XLASymNodeImpl to python because XLASymNodeImpl extends SymNode and SymNode has python bindings defined here :  https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/python/init.cpp#L1154 

`pySymNode->getPyObj` is a `PyObject*` for some `XLASymNodeImpl`. First we need to cast this `PyObject*` to `SymNode`  and we can do that because pybind11 automatically adds casters for any bound C++ classes : https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/python/init.cpp#L1154 

Now we finally have a c++ SymNode but it's still of the base class rather than the derived type (XLASymNodeImpl) that we need.

so we cast again `SymNode` to  `XLASymNodeImpl` so we can finally access `node()` and get the IR SizeNode out!

Reply:
Thanks for explaining.